### ISSUE TITLE
🐛 Bug Report: Premature close error

### ISSUE THREAD

**Issue by plague-doctor** [stonith404/pingvin-share#354]
### 👟 Reproduction steps

When downloading large file (couple of GB)  the download fails after few minutes. 
The docker throws:
```
[Nest] 47  - 12/08/2023, 12:47:07 PM   ERROR [ExpressAdapter] Premature close
Error [ERR_STREAM_PREMATURE_CLOSE]: Premature close
    at ServerResponse.onclose (node:internal/streams/end-of-stream:159:30)
    at ServerResponse.emit (node:events:526:35)
    at emitCloseNT (node:_http_server:1023:10)
    at Socket.onServerResponseClose (node:_http_server:278:5)
    at Socket.emit (node:events:526:35)
    at TCP.<anonymous> (node:net:337:12)
```
Sometimes this happens several times before the file if finally downloaded.

### 👍 Expected behavior

Successful download regardless the file size.

### 👎 Actual Behavior

As described above.

### 🌐 Browser

Firefox, Chromium, Brave

**Comment by stonith404** [comment]
That's strange I've never seen this error before. 

A few questions to debug this further:
Can you try it with another file? 
Does the error occur exactly at the same position (e.g at 2.3GB) ?

**Comment by plague-doctor** [comment]
Ok. I've tested it a little bit more. 
- no, it doesn't happen at the same position (I am not hitting any limits on reverse proxy either)
- it doesn't happen every time (it is rather an exception than a norm)
- it does happen with different files (but big ones - few GB)
- it happens with different external PCs

**Comment by stonith404** [comment]
Thanks for your detailed research. It's difficult to debug this if I can't reproduce it and I can't really find any useful information on the internet. I've created a Docker image that uses the `LTS` version of Nodejs, this might fix the issue. Could you try to use the image `stonith404/pingvin-share:development` temporarily?

**Comment by plague-doctor** [comment]
sure thing. Will do more tests with LTS

**Comment by plague-doctor** [comment]
A bit more investigation:
The LTS throws same errors:
```
[Nest] 41  - 12/14/2023, 12:53:24 AM   ERROR [ExpressAdapter] Premature close
Error [ERR_STREAM_PREMATURE_CLOSE]: Premature close
    at ServerResponse.onclose (node:internal/streams/end-of-stream:159:30)
    at ServerResponse.emit (node:events:526:35)
    at emitCloseNT (node:_http_server:1023:10)
    at Socket.onServerResponseClose (node:_http_server:278:5)
    at Socket.emit (node:events:526:35)
    at TCP.<anonymous> (node:net:337:12)
[Nest] 41  - 12/14/2023, 4:24:15 AM   ERROR [ExpressAdapter] Premature close
Error [ERR_STREAM_PREMATURE_CLOSE]: Premature close
    at ServerResponse.onclose (node:internal/streams/end-of-stream:159:30)
    at ServerResponse.emit (node:events:526:35)
    at emitCloseNT (node:_http_server:1023:10)
    at Socket.onServerResponseClose (node:_http_server:278:5)
    at Socket.emit (node:events:526:35)
    at TCP.<anonymous> (node:net:337:12)
```
However, I have tested different browsers this time.
- **Firefox**: Errors out with above. Sometimes after a few seconds, sometimes after a few minutes. Occasionally download is successful.
- **Librewolf**: Same as Firefox. (No mysteries here - it is Firefox too.)
- **Chromium**: No problems. 
- **Brave**: No problems.
All tests have been done from remote Linux machines running Linux Mint, Manjaro or Arch Linux.

**Comment by stonith404** [comment]
Hm okay, so it only occurs with Firefox. I'll try to investigate the further next week.

**Comment by csjoh** [comment]
Hello,

I'd like to add my 2c to this - I'm having the same problem, downloading an 8.8GB file from my own server. It goes well until it hits 1-1.2GB, then drops the download like a hot potato.

[Nest] 41  - 12/19/2023, 10:06:06 AM   ERROR [ExpressAdapter] Premature close
Error [ERR_STREAM_PREMATURE_CLOSE]: Premature close
    at ServerResponse.onclose (node:internal/streams/end-of-stream:159:30)
    at ServerResponse.emit (node:events:526:35)
    at emitCloseNT (node:_http_server:1023:10)
    at Socket.onServerResponseClose (node:_http_server:278:5)
    at Socket.emit (node:events:526:35)
    at TCP.<anonymous> (node:net:337:12)

Firefox on Linux Mint 21.2 and on Windows 10 22H2 both exhibit the same behaviour. I've not tried with any Chromium based browsers.

**Comment by stonith404** [comment]
Strange, I couldn't find anything useful on the internet. 

I'm not a Firefox user but did you experience this bug with other downloads from other websites?

**Comment by plague-doctor** [comment]
I doubt it would be related to Firefox. It is rather NestJS issue. Other projects reports similar problems...
Could we log some more to be able to pinpoint the issue?
https://github.com/nestjs/nest/issues/12283

**Comment by sebasdt** [comment]
Hey!

I may have found something while fiddling around.
The site doesn't timeout or cancel the download on Firefox while on HTTP, although the issue remains via HTTPS...
sadly it doesn't mean Pingvin container logs are fault-free, the same fault remains al be it less often.

The next thing I tested was accessing the site via HTTPS and the App URL is set to host ip and port.
Unfortunately, it resulted in the same fault where the download was canceled.

Still kinda pointing back to NestJs as @plague-doctor said. 

Anyway is there a workaround or known working version?

**Comment by stonith404** [comment]
If I have some time, I'll create a bug report in the NestJS repo with a small reproduction. Besides that I'm not really sure what I can do. 

Did the download of large files work before on Firefox?

**Comment by sebasdt** [comment]
> Did the download of large files work before on Firefox?

yes, it did. I can't download files bigger than 100mb while before a 2 GB file wasn't an issue.

**Comment by stonith404** [comment]
@sebasdt Can you remember approximately when downloading large files have stopped working?

**Comment by sebasdt** [comment]
The annoying part is I do not know at what version pingvin broke. (My backups don't go back that far and the last monthly december was just purged.)

The last known working state was back in 9 Dec, Although it could be later.

**Comment by axiomen** [comment]
This probably doesn't help much, but I tested it out of interest. I uploaded a 5GB file, then tested the following scenarions:

1. Download with Chrome -> no issue
2. Download with Firefox after downloading the same file with Chrome -> no issue
3. Download with Firefox only after closing Chrome -> Premature close (however, the download continues in case it's over unsecure HTTP)
4. Download with Firefox and Chrome simultaniously with Firefox starting first -> Premature close and fail to download on Chrome even over HTTP.
5. Resure download on Chrome after cancelling the simultanious download on Firefox -> no further issues.

I don't know if this helps in any way and if it's expected behaviour for this particular bug since I don't know how nestjs works, but thought I'd still post my findings.

**Comment by axiomen** [comment]
I've just tested older versions of pingvin-share, and I can tell that the issue starts happening with v0.21.1, while v0.21.0 seems to be working as expected.

**Comment by stonith404** [comment]
Thanks @axiomen for your observations!

Could you guys test version `v0.21.0` too? There aren't any db migrations since then so you only have to change the Docker image tag to `v0.21.0 `.

**Comment by csjoh** [comment]
I've tried all versions from and including 0.21.0 through latest and they all exhibit the same behaviour, i.e. dropping the download at around 1GB data transferred.

**Comment by axiomen** [comment]
That is strange. I have tested multiple large files, with single and multiple files per share and there was no issue on v0.21.0, while on v0.21.1 it happened immediatelly. I can re-test later today on another machine.

**Comment by sebasdt** [comment]
> I've tried all versions from and including 0.21.0 through latest and they all exhibit the same behaviour, i.e. dropping the download at around 1GB data transferred.

I've only tried 0.21.0 and am still only able to download on http although it with some errors.

Man, what is annoying is that we all have a common problem but the symptoms are all kinda different.

**Comment by csjoh** [comment]
Yeah.. I've poked at this for a bit to see if my environment was the culprit. I'm testing pingvin-share as a docker container on unRAID, and have it sitting behind an Nginx Proxy Manager reverse proxy which handles the TLS termination and passes the request through. NPM is also a docker container.

Thinking that was my issue, I spun up a Debian server and set up pingvin-share there, still routing it through NPM, and still the same problem. I then exposed the Debian host for pingvin-share directly on the internet - no dice.

I still saw the same errors in the logs and concluded that the environment is not the issue; something within the pingvin-share container must be wrong, but that's where my debugging skills end.

As I stated in my previous entry here, I've tried all versions from and including 0.21.0 through current, and they all do the same thing. I'll give 0.19.x a go and see if that has the same issues and report back.

**Comment by axiomen** [comment]
Tested on my second machine on v0.21.0, and the error is back. I have no idea how on my first machine the error consistently happens on v0.21.1+, but on v0.21.0 doesn't even after testing multiple scenarios with 1GB+ files. Both are containers spun on docker. Honestly at a loss here.

**Comment by stonith404** [comment]
Thank you guys for helping to debug this. It's frustrating as I can't really do much because it must be caused by the `StreamableFile` class of NestJS. And on the internet there is nothing that is related to this issue. I just upgraded all packages to the newest version, would you mind to test the `stonith404/pingvin-share:development` image?

If this doesn't work, I have to create an issue in the NestJS repository.

**Comment by axiomen** [comment]
No **Premature close error** on my first machine, but I did get another error upon finishing the download.

![image](https://github.com/stonith404/pingvin-share/assets/150123148/2ee06e29-477b-4c4e-97ac-80edb6097da6)

On my second machine, **Premature close error** is still present even on the dev image.

**Comment by csjoh** [comment]
On the development tag version, I was able to download one 3.5GB file and one 5.1GB file, but the zip file containing the two, with a total size of 8.8GB, failed after 1,138,822KB (1.13GB).

Retrying the first 3.5GB, it failed at 1.9GB:

[Nest] 41  - 01/11/2024, 2:35:23 PM   ERROR [StreamableFile] Premature close
Error [ERR_STREAM_PREMATURE_CLOSE]: Premature close
    at ServerResponse.onclose (node:internal/streams/end-of-stream:159:30)
    at ServerResponse.emit (node:events:526:35)
    at emitCloseNT (node:_http_server:1023:10)
    at Socket.onServerResponseClose (node:_http_server:278:5)
    at Socket.emit (node:events:526:35)
    at TCP.<anonymous> (node:net:337:12)

I've noticed a few other things as well, but I'll open separate tickets for those.

**Comment by stonith404** [comment]
Okay thanks. I think the only option now is to report it to NestJS.

**Comment by axiomen** [comment]
I did some more testing, and noticed that on Chrome the error happens if multiple simulatneous downloads from Pingvin Share are running. For instance, if I'm downloading three files from the same share (file1, file2, and zip of file1 and file2), I'll get **Premature close** error, but not all downloads are cut off (file1 and file2 stop downloading, while the zip continues to download).

**Comment by stonith404** [comment]
So I debugged this issue a bit further and I found out that the backend logs `Premature close error` if the the browser cancels the download. My guess is that the issue occurs when the network is unstable and the client devices gets disconnected for a short time. 

For example I'm using a VPN and I was able to reproduce this issue with the VPN but if I disable the VPN the download works. My VPN isn't really reliable because sometimes if I visit a website I get the error "Network has changed".

Could this be the issue or are you guys sure that your connection on the server and client are stable?

**Comment by fosq** [comment]
I tested it with 2 friends, both with fiber connection and near 5km radius. I'm using pingvin-share behind SWAG reverse-proxy. `Download All` button seems to throw Premature Close error more often on a share with multiple files, than downloading the files separately. Downloading single files have been giving variable success rate, sometimes it downloads it fully, sometimes it fails the file download with Premature close error.

**Comment by stonith404** [comment]
@fosq Thanks for taking the time for helping to debug this. Would you mind to test if the issue also occurs when you download the file over your local network, without a proxy in front?

**Comment by stonith404** [comment]
So, I just rewrote the download endpoint completely and replaced the `StreamableFile` with the `download` method of Express. I've tested the download with a VPN, over the local network, over the internet, and concurrent downloads. Would you mind testing the `stonith404/pingvin-share:development` image again? 🙏

**Comment by axiomen** [comment]
Tested on two hosts. Unfortunately, still the same probelm.

Host #1: remote over HTTPS, three simultaneous downloads, all failed at a certain point.

![host1](https://github.com/stonith404/pingvin-share/assets/150123148/54b64147-52c6-460c-ba79-43af80151656)

Host #2: local network over HTTP, three (different) simultaneous downloads, all but one failed.

![host2](https://github.com/stonith404/pingvin-share/assets/150123148/8174de47-f27e-4461-9231-0d38db74879d)

Tested on Firefox.

**Comment by stonith404** [comment]
@axiomen Thanks for your research. I made some researches too and it's extremely strange. You're right that the issue still occurs because I only tested it with a chromium based browser. Downloading files in a chromium based browser concurrently works with no problems for me. Downloading one file at the same time in Firefox works for me too, but not multiple files at the same time. If I start the download in Chrome first and then start the download in Firefox, the Firefox download fails, but if I start the download in Firefox first the files get downloaded successfully in both browsers.

> it seems that this is more of an Nginx issue

You said that you tested the download over the local network, was there Nginx in front too?

**Comment by axiomen** [comment]
Sorry, I was in a rush, so I didn't elaborate on my comment. From my understanding, Docker image comes with bundled Nginx, correct? If not, disregard that.

This is what I've noticed the process usually goes like: Firefox shows the download is still in progress, even though the transfer shows no activity anymore (i.e. 0 MB/s) -> Network Monitor reports "Blocked" -> Pingvin share throws a "Request aborted" / "Premature close" error.

**Comment by axiomen** [comment]
Okay, further investigation has led me to something interesting (again, might not be relted at all). There is a mismatch between reported file size on Pingvin Share interface vs the one that the browsers report (valid for both Chrome and Firefox).

![Screenshot 2024-01-26 093026](https://github.com/stonith404/pingvin-share/assets/150123148/c56aef86-1cf1-41f1-ad52-ca4422e3ca28)

Which has led me to [this issue](https://drupal.stackexchange.com/questions/273562/firefox-cant-download-private-pdfs). Relevant parts:

> I've looked at the response headers when the file's downloading and all is in order, until I get this NS_ERROR_NET_PARTIAL_TRANSFER message and it fails.

> I have an issue with FireFox where it can't download generated private file PDFs, yet Chrome has no such issue. 

> The real issue is the mismatch between the file's real size on the server and the size Drupal stores in the file_managed table, filesize column.

I did check the content-length in the HTTP Headers when the transferred fails, which reports the correct size, so I'm not really sure where the mismatch happens.

**Comment by axiomen** [comment]
@stonith404 I got it working, but feel like it's more of a workaround?

Mounted nginx.conf to the host and added "proxy_buffering off", so it looks like this:

```
location /api {
            proxy_buffering off;
            proxy_pass http://localhost:8080;
            proxy_set_header X-Forwarded-Host $host:$server_port;
            proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
            proxy_set_header X-Forwarded-Proto $scheme;
        }
```

I'm not sure what exactly is going on, and from my understanding this is not the best practice, but it works now.

**Comment by sebasdt** [comment]
> Docker image comes with bundled Nginx, correct?

It seems like it yes. Even the bare install without docker uses Nginx.
Maybe it's possible to temporarily replace Nginx with Caddy or Apache to see if Nginx is being a dufus?

**Comment by stonith404** [comment]
I just replaced Nginx with Caddy in the latest development image. For me the error strangely still occurs if I disable proxy buffering or use Caddy. Would you mind to test the image again?

@axiomen Is the error completely gone if you disable proxy buffering?

**Comment by axiomen** [comment]
I managed to do a little bit of testing over HTTPS (latest image, slightly adjusted nginx on my host), but on my phone with Firefox Android using mobile data. Ran one test with two concurrent downloads with no errors. I'm starting to think there's something up with the firewall on the network I've done all (or most of) my testings on so far. Once the downloads fail, I noticed I cannot access anything on my domain on this network for a while, however everything remains accessible through my phone data.

**Comment by axiomen** [comment]
I did a quick test, and it's not only on Firefox under the aformentioned conditions. Chrome is also affected. 

A rather strange thing I've noticed is my domain becomes unreachable for a minute or so after the downloads fail. Containers keep running, so does swag (nginx), and there are no error logs indicating something's wrong. For good measure, I tested file transfer over my Nextcloud instance (in case there's some domain hosting issue), and that worked fine.

**Comment by sebasdt** [comment]
Hi!

Well I just gave the latest dev build a look, It seems like I got a different error.. 
That might just be because im on a dev built.

**Comment by stonith404** [comment]
Thanks @axiomen and @sebasdt. This issue is really difficult the fix, because I'm still not sure where the problem is. But I'll keep you up to date.

**Comment by axiomen** [comment]
I looked a bit more through the _swag_ logs, including _fail2ban_ module, and also _Network_ stats in Firefox, and this popped out for me.

![Screenshot 2024-02-06 094704](https://github.com/stonith404/pingvin-share/assets/150123148/31ca42fd-5883-4e38-8698-e84670359ab6)

As you can see, I'm getting quite a few 401 errors due to "Slow server response time" while downloading, which ultimately ends with fail2ban temporarily blocking the IP address, and failing the transfer.

**Comment by stonith404** [comment]
@axiomen Thanks for participating in #401 let's discuss related things there, instead of here.

Did disabling the fail2ban jail really fix the issue? I'm not using fail2ban and the issue still occurs. I still have no clue where this issue occurs as the symptoms are different for every user. For example you said that the issue also occurs in Chrome but I made many tests and it never failed for me in Chrome.

**Comment by axiomen** [comment]
@stonith404 I've run multiple tests, upload + download of two 4-5 GB files, and all tests passed. That's on the dev image with Caddy. As long as I disable `nginx-unauthorized` fail2ban jail OR log in while downloading, the dev image works as expected.

---

I did stumble on another issue with this image though, where editing a share and attempting to add a file results in an upload error and the share disappearing completely on the web interface. The share is still present on the disk, however. This is the error:

```
[Nest] 49  - 02/12/2024, 10:08:38 AM   ERROR [ExceptionsHandler] request entity too large
PayloadTooLargeError: request entity too large
    at readStream (/opt/app/backend/node_modules/raw-body/index.js:163:17)
    at getRawBody (/opt/app/backend/node_modules/raw-body/index.js:116:12)
    at read (/opt/app/backend/node_modules/body-parser/lib/read.js:79:3)
    at rawParser (/opt/app/backend/node_modules/body-parser/lib/types/raw.js:81:5)
    at Layer.handle [as handle_request] (/opt/app/backend/node_modules/express/lib/router/layer.js:95:5)
    at trim_prefix (/opt/app/backend/node_modules/express/lib/router/index.js:328:13)
    at /opt/app/backend/node_modules/express/lib/router/index.js:286:9
    at Function.process_params (/opt/app/backend/node_modules/express/lib/router/index.js:346:12)
    at next (/opt/app/backend/node_modules/express/lib/router/index.js:280:10)
    at expressInit (/opt/app/backend/node_modules/express/lib/middleware/init.js:40:5)
```

**Comment by stonith404** [comment]
@axiomen Okay thanks for your feedback. I'll test it ASAP.

**Comment by stonith404** [comment]
@axiomen I don't have fail2ban and the error still occurs when I'm signed in :/ Did you maybe change something else?

Regarding the `PayloadTooLargeError`, this is fortunately easy to fix and only caused on the dev image as I have increased the chunk size for testing purposes.

**Comment by axiomen** [comment]
That's really bizarre. I haven't changed anything else. This is my nginx configuration file, if it helps in any way:

```
## Version 2023/05/31
# make sure that your pingvin-share container is named pingvin-share
# make sure that your dns has a cname set for pingvin-share

server {
    listen 443 ssl http2;
    listen [::]:443 ssl http2;

    server_name REDACTED.*;

    include /config/nginx/ssl.conf;

    client_max_body_size 0;

    location / {
        include /config/nginx/proxy.conf;
        include /config/nginx/resolver.conf;
        set $upstream_app REDACTED;
        set $upstream_port 3000;
        set $upstream_proto http;
        proxy_pass $upstream_proto://$upstream_app:$upstream_port;

    }

}
```

**Comment by stonith404** [comment]
The bug #401 should be resolved in the most recent 'development' image. Would you mind to try if you still don't encounter the "premature error" anymore?

**Comment by axiomen** [comment]
@stonith404 I'm getting `502 Bad Gateway` with the latest dev image.

**Comment by earthly-alien** [comment]
Deleted and redownload the image, it's working now. I'll test for the 401 error as soon as I get to my PC.

**Comment by axiomen** [comment]
@stonith404 removed and pulled the newest dev image, tested a 24GB file transfer, and haven't seen any errors. Neither the 401, nor the Premature.

**Comment by babsz02** [comment]
@stonith404 Thanks for this quick version
I've just tested it with an 8GB archive, several 2GB video files and no crashes.
But still a very high use of RAM and even the transfer finished downloading I keep this use of RAM.
I have to reboot my VM to get it back to normal.
This happens to me on upload and download.

<img width="599" alt="Capture d’écran 2024-02-27 à 23 31 21" src="https://github.com/stonith404/pingvin-share/assets/45185204/d36c8fff-352b-45a4-8f0f-60e75c1a4ed3">

**Comment by axiomen** [comment]
I'm guessing whatever you run your VM on can't tell apart free vs cached memory, so it reports that most of your memory is taken, while in reality 84% is cached (as shown with the free command). The cached part is probably the file you're downloading cached directly in the memory.

**Comment by stonith404** [comment]
This should be fixed in `v0.22.2`. Feel free to reopen the issue if you encounter the bug again.

**Comment by stonith404** [comment]
I'm reopening this as @axiomen has reported that the issue still exists.

**Comment by tranxuanthin** [comment]
@stonith404 When implementing streaming of files with experienced users (getting a ranged file) from Minio using the S3 SDK client and obtaining a readable stream, I utilize 'StreamableFile' to pipe the client. Initially, it works fine. However, when I attempt to tassel the video, an error "ERROR [StreamableFile] Premature close" occurs. I've tested this multiple times, and the error consistently appears, indicating a potential interruption in the stream.

**Comment by stonith404** [comment]
This bug is really strange. @axiomen has reported that it doesn't occur anymore so I'll close it again... If you notice the error again please comment here.

**Comment by sebasdt** [comment]
It's just like a herobrine you try to fix It/remove the bug and it just pops up again.😜

**Comment by drustan** [comment]
@axiomen  On a remote network, behind an Nginx reverse proxy

**Comment by drustan** [comment]
The weird thing is that it's always the first try that fails, and on the second attempt, it works just fine.

**Comment by hasibkyau** [comment]
I am facing the same issue, and in my case, I was trying to stream pdf files. I tested with different file size and found the Premature close issue only for files whose size is above 5MB. I tested on my Windows machine along with Edge, Firefox and Chromium browsers. Followings are the test cases:

Test Case 01:
File type: PDF, File Size: 1MB
Result: Pass.

Test Case 02:
File type: PDF, File Size: 5MB
Result: Pass.

Test Case 03:
File type: PDF, File Size: 10MB
Result: Failed. ( ERROR [StreamableFile] Premature close)

Test Case 04:
File type: PDF, File SIze: 7MB
Result: Failed. ( ERROR [StreamableFile] Premature close)

**Comment by MattyMay** [comment]
Also having this issue. Just spent a couple hours thinking it must be the config of my reverse proxy in front of Pingvin, but now I'm not so sure.

Some details:
- 500MB junk file download fails 100% of the time with:
    - Firefox
- 500MB junk file download succeeds 100% of the time with:
    - Chrome
    - `curl`
    - `wget`
- Internet connection is definitely a bit flaky, but nothing that's noticeable "during normal usage"
- Cloudflare DNS with requests proxied through Cloudflare
- NGINX reverse proxy sitting in front of Pingvin. Config:
```
server {
    listen 443 ssl http2;
    server_name [REDACTED];

    include /etc/nginx/conf.d/snippets/ssl.conf;

    set $upstream http://${PINGVIN_HOST}:${PINGVIN_PORT};

    client_max_body_size      0;
    proxy_buffering           off;
    proxy_request_buffering   off;
    sendfile                  off;

    location / {
        proxy_set_header     Host $host;
        proxy_set_header     X-Real-IP $remote_addr;
        proxy_set_header     X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header     X-Forwarded-Proto $scheme;
        proxy_pass           $upstream;
    }

    location /robots.txt {
        add_header Content-Type text/plain;
        return 200 "User-agent: *\nDisallow: /\n";
    }
}

```

- The error is as follows:
```
pingvin  | {"level":"error","ts":1720589993.0274456,"logger":"http.handlers.reverse_proxy","msg":"aborting with incomplete response","upstream":"localhost:8080","duration":0.046721392,"request":{"remote_ip":"[REDACTED]","remote_port":"39394","client_ip":"[REDACTED]","proto":"HTTP/1.0","method":"GET","host":"[REDACTED]","uri":"/api/shares/IwOTM2O/files/6b2d8478-7eba-43ef-8cbb-08a3759c75a0","headers":{"X-Real-Ip":["[REDACTED]"],"Accept-Encoding":["gzip, br"],"Accept-Language":["en-US,en;q=0.5"],"Cdn-Loop":["cloudflare"],"Upgrade-Insecure-Requests":["1"],"X-Forwarded-For":["[REDACTED]"],"Sec-Fetch-Dest":["empty"],"Cf-Ray":["8a0e35c4cd811000-LAX"],"Sec-Fetch-Site":["same-origin"],"Cf-Connecting-Ip":["[REDACTED]"],"Sec-Fetch-Mode":["same-origin"],"Priority":["u=4"],"X-Forwarded-Host":["[REDACTED]:{server_port}"],"User-Agent":["Mozilla/5.0 (X11; Linux x86_64; rv:127.0) Gecko/20100101 Firefox/127.0"],"Referer":["[REDACTED]"],"Cf-Ipcountry":["US"],"Cookie":[],"X-Forwarded-Proto":["http"],"Accept":["text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,*/*;q=0.8"],"Cf-Visitor":["{\"scheme\":\"https\"}"]}},"error":"writing: write tcp [REDACTED]->[REDACTED]: write: connection reset by peer"}
pingvin  | [Nest] 38  - 07/10/2024, 5:39:53 AM   ERROR [StreamableFile] Premature close
pingvin  | Error [ERR_STREAM_PREMATURE_CLOSE]: Premature close
pingvin  |     at ServerResponse.onclose (node:internal/streams/end-of-stream:159:30)
pingvin  |     at ServerResponse.emit (node:events:531:35)
pingvin  |     at emitCloseNT (node:_http_server:1020:10)
pingvin  |     at Socket.onServerResponseClose (node:_http_server:278:5)
pingvin  |     at Socket.emit (node:events:531:35)
pingvin  |     at TCP.<anonymous> (node:net:338:12)
```

**Comment by mossroy** [comment]
Same issue here with v1.2.0.
Downloads stop before 1GB with Firefox 131.0.2, but work with Chromium 129.

Error log on server side:
```
[Nest] 48  - 10/16/2024, 5:18:19 PM   ERROR [StreamableFile] Premature close
Error [ERR_STREAM_PREMATURE_CLOSE]: Premature close
    at ServerResponse.onclose (node:internal/streams/end-of-stream:159:30)
    at ServerResponse.emit (node:events:531:35)
    at emitCloseNT (node:_http_server:1021:10)
    at Socket.onServerResponseClose (node:_http_server:277:5)
    at Socket.emit (node:events:531:35)
    at TCP.<anonymous> (node:net:339:12)
```

In my case, the bytes go through a Traefik reverse-proxy.
Everything on my local (and reliable) network

**Comment by babsz02** [comment]
Hello, some time ago I had again the problem with the last version of Pingvin and thanks to ChatGPT I modified my proxy.conf file via Swag with what he suggested and it's been working perfectly ever since.

# Désactiver le buffering proxy
    proxy_buffering off;
    proxy_request_buffering off;
    
    # Désactiver la redirection proxy
    proxy_redirect off;

    # Timeout augmenté pour les fichiers volumineux
    proxy_connect_timeout 600;
    proxy_send_timeout 600;
    proxy_read_timeout 600;
    send_timeout 600;

**Comment by mossroy** [comment]
@babsz02 you imply that it might come from the behavior of your frontal reverse-proxy?

In any case, I'd be happy to help debugging this issue if I can (I won't be able to help on the typescript code, unfortunately)

**Comment by babsz02** [comment]
Indeed, it comes from my riverse proxy in my case Swag proxy.
Since I made the changes mentioned above, everything's working again.

**Comment by earthly-alien** [comment]
@babsz02 If you scroll through the comments, you'll see that I reproduced the error on an instance without a reverse proxy, so I don't think it's that in every case.

**Comment by earthly-alien** [comment]
@babsz02 if you scroll through the comments, you'll see the issue has been reproduced even on instances that aren't behind a reverse proxy, so I don't think that's really it in all cases.

**Comment by sfreek76** [comment]
Hello

I have the same issue, I'm behind a reverse Proxy from a Synology NAS and it stops downloading with Firefox after a few GB.

    level=errorts=1737046614.493012logger=http.handlers.reverse_proxymsg=aborting with incomplete responseupstream=localhost:8080duration=0.022685417request.client_ip=172.22.0.1request.headers.Accept=["text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8"]request.headers.Accept-Encoding=["gzip, deflate, br, zstd"]request.headers.Accept-Language=["de,en-US;q=0.7,en;q=0.3"]request.headers.Cookie=["REDACTED"]request.headers.Dnt=["1"]request.headers.Priority=["u=2"]request.headers.Referer=["https://myDomain.com/share/2EzHmjN8"]request.headers.Sec-Fetch-Dest=["empty"]request.headers.Sec-Fetch-Mode=["same-origin"]request.headers.Sec-Fetch-Site=["same-origin"]request.headers.Sec-Gpc=["1"]request.headers.Upgrade-Insecure-Requests=["1"]request.headers.User-Agent=["Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:134.0) Gecko/20100101 Firefox/134.0"]request.headers.X-Forwarded-For=["myIP, 172.22.0.1"]request.headers.X-Forwarded-Host=["myDomain.com"]request.headers.X-Forwarded-Proto=["https"]request.headers.X-Real-Ip=["myIP"]request.host=myDomain.com.comrequest.method=GETrequest.proto=HTTP/1.1request.remote_ip=172.22.0.1request.remote_port=63698request.uri=/api/shares/2EzHmjN8/files/ziperror=writing: write tcp 172.22.0.2:3000->172.22.0.1:63698: write: broken pipe

[Nest] 51  - 01/16/2025, 4:56:54 PM   ERROR [StreamableFile] Premature close
Error [ERR_STREAM_PREMATURE_CLOSE]: Premature close
    at ServerResponse.onclose (node:internal/streams/end-of-stream:159:30)
    at ServerResponse.emit (node:events:530:35)
    at emitCloseNT (node:_http_server:1021:10)
    at Socket.onServerResponseClose (node:_http_server:277:5)
    at Socket.emit (node:events:530:35)

        at TCP.<anonymous> (node:net:343:12)

**Comment by coenstam-united** [comment]
Hi,

First of all, thanks for this project. I am experiencing the same issue in Firefox, Chrome, and Brave, as well as with smaller files around 1GB. I have read all the above comments, and it seems like this isn't a browser-specific issue nor a reverse proxy issue. Hopefully, this bug can be resolved soon, as it makes Pingvin unreliable for me.

**Comment by Archer36** [comment]
Hi all,

Recently tried out Pingvin and looking to use it to share 5+ GB action camera video files. I too am encountering the Premature close on Firefox, especially when downloading multiple files at the same time. So far I've tried the below on my local LAN:

1. Pingvin with no HTTPS reverse proxy: seems to work
2. Pingvin with nginx HTTPS reverse proxy: fails
3. Pingvin with nginx HTTPS reverse proxy with babsz02 disabling buffering tweaks: fails
4. Pingvin with Caddy HTTPS reverse proxy: seems to go a little longer but then fails

I love the application so far and would love to see this one fixed so I can still use it. Let me know if I can provide any logs or additional tests.
